crates/db_views/src/post_view.rs

```diff
@@ -417,12 +429,43 @@ impl PostView {
   }
 }
 
-#[derive(Default)]
+impl PaginationCursor {
+  // get cursor for page after the given posts
+  pub fn after(posts: &[PostView]) -> Option<PaginationCursor> {
+    posts
+      .last()
+      .map(|p| PaginationCursor(format!("P{:x}", p.post.id.0)))
```

Comment:
This impl should have `to_id`, and `from_id` functions. Below you're converting from radix, but here you're just using the post id.

That tests are passing with this issue, tells me we need to test pagination either here, or in `api_tests/`.

Reply:
> Below you're converting from radix, but here you're just using the post id.

I am converting to radix 16 (that's what the `{:x}` does), so the code is correct. I did test the basic functionality of this manually a bit, though I agree adding tests would probably be a good idea (just lazy).

The reason I added these more "indirect" functions (after + read) is that I'm not sure if actually using post id to represent the token is the best idea. `from_id` would be somewhat less expressive because it doesn't tell you what it means - what does it mean to get a pagination token from one post id? That's why I am passing one page of posts into the function, and the function clearly gives you a token that represents the page `after` the given input page.

A `to_id` function is also not really used except internally by the `read()` function.

Reply:
I've replaced the function with `after_post(PostView)` and moved the part that checks for page emptyness to the outside function

Reply:
>  I'm not sure if actually using post id to represent the token is the best idea.

That's essentially what Reddit does with paging. If you are paging say your blended-subreddit list of communities, it passes a enhanced variation of the id= of the post to the next page (and even part of the front-end URL convention). This is what the 2nd-page of Reddit post list  looks like on URL: https://old.reddit.com/?count=25&after=t3_15y74bo

Instead of integer post id numbers, they use a generated pattern of letters and numbers. Example: https://old.reddit.com/comments/15y74bo/

Reply:
I mean it's pretty clean, it just requires an extra db fetch that we could in theory avoid

Reply:
Since the cursor is actually a compound of columns then, a `to_cursor` and `from_cursor`, that can decode / encode the object / tuple might be best.

Reply:
The issue with that is that either it would mean the pagination cursor would be huge (the whole post aggregates object), or there would need to be many special cases for every kind of sort (since every sort needs a different subset of information for the pagination column). it would also require some compact url-safe encoding (json+base64 would work but would again result in huge tokens)

Reply:
i'll resolve this because it's now the same discussion as https://github.com/LemmyNet/lemmy/pull/3872#discussion_r1301858017

Reply:
> But that's also the reason i made the token into a visibly opaque string so we can change this freely later (no breaking change required)

It just have to have the sort columns, right?